meteor/diffmaps.py

```diff
@@ -0,0 +1,195 @@
+) -> rs.DataSet | None:
+    """
+    Compute k-weighted differences between native and derivative amplitudes and phases.
+
+    Assumes that scaling has already been applied to the amplitudes before calling this function.
+
+    Parameters:
+    -----------
+    dataset : rs.DataSet
+        The input dataset containing columns for native and derivative amplitudes/phases.
+    native_amplitudes : str
+        Column label for native amplitudes in the dataset.
+    derivative_amplitudes : str
+        Column label for derivative amplitudes in the dataset.
+    native_phases : str, optional
+        Column label for native phases.
+    derivative_phases : str, optional
+        Column label for derivative phases, by default None.
+    sigf_native : str
+        Column label for uncertainties of native amplitudes.
+    sigf_deriv : str
+        Column label for uncertainties of derivative amplitudes.
+    kweight : float, optional
+        k-weight factor, by default 1.0.
+    optimize_kweight : bool, optional
+        Whether to optimize the kweight using negentropy, by default False.
+    inplace : bool, optional
+        Whether to modify the dataset in place. Default is False.
+
+    Returns:
+    --------
+    rs.DataSet | None
+        The modified dataset with k-weighted differences, if inplace=False, otherwise None.
+    """
+    if not inplace:
+        dataset = dataset.copy()
+
+    # Compute differences between native and derivative amplitudes and phases
+    dataset = compute_deltafofo(
+        dataset=dataset,
+        native_amplitudes=native_amplitudes,
+        derivative_amplitudes=derivative_amplitudes,
+        native_phases=native_phases,
+        derivative_phases=derivative_phases,
+        inplace=inplace,
+    )
+
+    delta_amplitudes = dataset["DeltaFoFo"]
+    sigdelta_amplitudes = np.sqrt(dataset[sigf_deriv] ** 2 + dataset[sigf_native] ** 2)
+
+    if optimize_kweight:
+
+        def negentropy_objective(kweight_value: float) -> float:
+            # Apply k-weighting to DeltaFoFo
+            weights = compute_kweights(
+                delta_amplitudes, sigdelta_amplitudes, kweight_value
+            )
+            weighted_delta_fofo = delta_amplitudes * weights
+            dataset["DeltaFoFoKWeighted"] = weighted_delta_fofo
+
+            # Convert weighted amplitudes and phases to a map
+            delta_map = compute_map_from_coefficients(
+                map_coefficients=dataset,
+                amplitude_label="DeltaFoFoKWeighted",
+                phase_label=native_phases,
+                map_sampling=TV_MAP_SAMPLING,
+            )
+
+            delta_map_as_array = np.array(delta_map.grid)
+
+            # Compute negentropy of the map
+            return negentropy(delta_map_as_array.flatten())
+
+        # Optimize kweight using negentropy objective
+        maximizer = ScalarMaximizer(objective=negentropy_objective)
+        maximizer.optimize_over_explicit_values(arguments_to_scan=(0.0, 1.0))
```

Comment:
I don't think this is what you want to do -- this will just test 0.0 and 1.0 (only!)